oauth2/src/main/java/com/linecorp/armeria/internal/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -0,0 +1,224 @@
+     * @param endpointPath A URI path that corresponds to the Authorization endpoint of the
+     *                     OAuth 2 system.
+     * @param clientAuthorization Provides client authorization for the OAuth requests,
+     *                            as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected AbstractOAuth2Request(WebClient endpoint, String endpointPath,
+                                    @Nullable ClientAuthorization clientAuthorization) {
+        this.endpoint = Objects.requireNonNull(endpoint, "endpoint");
+        this.endpointPath = Objects.requireNonNull(endpointPath, "endpointPath");
+        this.clientAuthorization = clientAuthorization; // optional
+    }
+
+    /**
+     * Returns the {@link WebClient} of the authorization endpoint.
+     */
+    protected final WebClient endpoint() {
+        return endpoint;
+    }
+
+    /**
+     * Returns the authorization endpoint path.
+     */
+    protected final String endpointPath() {
+        return endpointPath;
+    }
+
+    /**
+     * Returns the client authorization object.
+     */
+    @Nullable
+    protected final ClientAuthorization clientAuthorization() {
+        return clientAuthorization;
+    }
+
+    /**
+     * Extracts data from OK response and converts it to the target type {@code T}.
+     */
+    protected abstract T extractOkResults(AggregatedHttpResponse response, QueryParams requestFormData);
+
+    /**
+     * Returns the value for the {@link HttpHeaderNames#AUTHORIZATION}.
+     */
+    @Nullable
+    protected final String authorizationHeaderValue() {
+        return clientAuthorization == null ? null : clientAuthorization.asHeaderValue();
+    }
+
+    /**
+     * Sets client credentials as form data parameters,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected final void addCredentialsAsBodyParameters(QueryParamsBuilder formBuilder) {
+        if (clientAuthorization != null) {
+            clientAuthorization.addAsBodyParameters(formBuilder);
+        }
+    }
+
+    /**
+     * Makes a request to the authorization endpoint using supplied {@code requestForm} parameters and converts
+     * the result to the given type {@code T}.
+     */
+    protected CompletableFuture<T> executeWithParameters(QueryParams requestFormData) {
+        final HttpResponse response = endpoint().execute(createHttpRequest(endpointPath, requestFormData));
+        // when response aggregated, then extract the results...
+        return response.aggregate().thenApply(r -> extractResults(r, requestFormData));
+    }
+
+    /**
+     * Produces {@link HttpRequest} based on this object.
+     */
+    private HttpRequest createHttpRequest(String endpointPath, QueryParams requestFormData) {
+        final RequestHeadersBuilder headersBuilder =
+                RequestHeaders.of(HttpMethod.POST, endpointPath).toBuilder();
```

Comment:
We can directly create a `RequestHeadersBuilder`
```suggestion
        final RequestHeadersBuilder headersBuilder =
                RequestHeaders.builder(HttpMethod.POST, endpointPath);
```

Reply:
Right. Fixed